텔러가 카드 선택하지 않았을 때 랜덤으로 선택된 값 보내주는 기능

sbyeol3 opened the pull request:
## 💁 설명

- 일단 카드 토픽을 담아두는 건 json으로 분리를 했습니다. 귀차나서 몇 개 안했는데 추후 심심할 때 추가해봅시다..
- 라운드가 시작되는 건 항상 텔러가 선택되는 게 시작이므로 거기서 1분이 지나도 state가 텔러이면 강제로 선택 `forceTellerSelect`
- 아직 연결은 안했지만 누가 뼈대를 잡아둔게 있어서 `forceGuesserSelect`도 수정했습니다.
- 사실 `forceGuesserSelect` 코드가 좀 더럽긴 한데.. 사실 최대 인원이 6명이라 시간복잡도는 고만고만할듯 하긴 해요. ㅜ
- 하지만 리팩토링 할 것입니다.. 🤟
- `selectCardFromUser` 로직이 중복으로 사용돼서 아예 함수로 뺐습니다!
- 나중에는 시간도 같이 보내주도록 수정하면 좋을 듯 해요~

### GIF 참고
> 데브 상 5초동안 선택안하면 하게 했지만 push한 코드에는 60000 밀리세컨즈임다
![화면 기록 2020-12-09 오전 12 32 58 mov](https://user-images.githubusercontent.com/43198553/101504599-bf39cf00-39b6-11eb-99d6-dfb8bc47b588.gif)

## 📑 체크리스트

> 구현한 목록 체크리스트

- [ ] 텔러가 1분동안 선택하지 않으면 모두에게 토픽이 전달되고 텔러에게는 카드 아이디도 같이 줍니다.


## 🚧 주의 사항

> PR을 읽을 때 살펴볼 사항

### 리팩토링? 모듈화?
`socketIO`로 emit 하는 함수는 `@utils/socket`으로 분리했어요.
- users 키가 있으면 users 순회하면서 모두 보내주고
- 아니면 socketID에게만 보내줍니다.

### 아마 아직 토픽 구현이 다 안되어서 undefined로 갈 확률이 큼

close #154

sbyeol3 committed:
Feat : card Topic 담아두는 json 추가 (#154)

- 일단 귀차나서 17개만 씀..

sbyeol3 committed:
Feat : 매직넘버 분리 및 emit 유틸함수 추가

- emit 하는 함수 유틸로 분리
- 매직넘버 분리

sbyeol3 committed:
Feat : game 클래스 함수 수정 + game state 상수 수정 (#154)

- 강제로 텔러 카드 선택하는 함수 추가
- 강제러 guesser 카드 선택하는 함수 수정
- 모듈화도 해봄

sbyeol3 committed:
Refactor : 내 아이디 빼고 넘겨주도록 추가

- other guesser decision은 나 빼고 보내야 하므로 ... filter 추가

ramram1048 commented on src/backend/utils/cardTopic.json:
```diff
@@ -0,0 +1,19 @@
+{
+    "1": "거인",
+    "2": "바라보다",
```
이거 정말 큰일이네요. 200장이 넘는걸 다 할수는 없을 것 같고 저희가 카드 상한을 nginx에 담긴 카드보다 적게 잡아야되겠어요...

Front-line-dev replied:
꼭 정확하게 적어야 할 필요는 없으니 분담하면 10분내에 끝낼 수 있지 않을까요? 그냥 제 생각입니다

sbyeol3 replied:
저도 민성님처럼 생각!ㅋㅋㅋ

ramram1048 commented on src/backend/utils/number.js:
```diff
@@ -1,5 +1,5 @@
 export const PLAYER = {
-  MIN: 3,
+  MIN: 2,
```
이건 디버깅하려고 임시로 바꾸신건가요?

sbyeol3 replied:
뺐다고 생각했는데 들어가버렸습니다 고치겟ㅅ읍니다

ramram1048 reviewed (commented):
백엔드 Game이 점점 뚱뚱해지는게 걱정이네요. 어쩌면 게임 흐름(scene)과 관련된 친구들은 다른 파일로 함수를 옮겨버리고, Game에서 걔네를 bind(this)로 연결하면 줄일 수 있을지도 모르겠다는 생각이 들었습니다. 생각만 들었습니다.

Front-line-dev commented on src/backend/game/Game.js:
```diff
@@ -100,23 +108,49 @@ export default class Game {
     return [...this.users.values()];
   }
 
+  selectCardFromUser({ socketID, teller = true }) {
+    const user = this.users.get(socketID);
+    const cardID = generateRandom.pickOneFromArray(user.cards);
+    user.submittedCard = cardID;
+    return teller ? { cardID, topic: TOPIC[cardID] } : { cardID };
+  }
```
게임 보다는 User에 만드는게 그나마 Game 크기를 줄일 수 있을거 같네요

Front-line-dev commented on src/backend/game/Game.js:
```diff
@@ -100,23 +108,49 @@ export default class Game {
+    return teller ? { cardID, topic: TOPIC[cardID] } : { cardID };
+  }
+
+  forceTellerSelect(tellerID) {
+    const { cardID, topic } = this.selectCardFromUser({
+      socketID: tellerID,
+      teller: true,
+    });
+    this.updateTopic(topic);
+    this.users.forEach((user) => {
+      const { socketID } = user;
+      const isTeller = socketID === tellerID;
+      const name = isTeller ? 'teller select card' : 'teller decision';
+      const params = isTeller ? { cardID, topic } : { cardID };
+      emit({ socketID, name, params });
+    });
+    this.updateState(GAME_STATE.GUESSER);
+  }
```
요 함수는 socket 쪽에서 작성하면 좋을거 같아요

Front-line-dev commented on src/backend/utils/socket.js:
```diff
@@ -1,3 +1,11 @@
+  if (users && !socketID) {
+    users.forEach((user) => socketIO.to(user.socketID).emit(name, params));
+  } else socketIO.to(socketID).emit(name, params);
+};
```
음 저는 이렇게 utils/socket 보다는 Game의 로직을 socket/~~~ 로 옮기는게 좋을거 같아요.

Front-line-dev commented on src/backend/utils/cardTopic.json:
```diff
@@ -0,0 +1,19 @@
+    "15": "아슬아슬",
+    "16": "변장",
+    "17": "지루함"
+}
```
언젠간, 누구가는 해야 할 일이였는데 이렇게 스타트를 끊어주시다니 역시 행동력이 엄청나십니다

Front-line-dev commented on src/backend/utils/gameState.js:
```diff
@@ -2,9 +2,10 @@ const GAME_STATE = {
-  VOTE: 3,
-  RESULT: 4,
-  END: 5,
+  DISCUSSION: 3,
```
앗 제가 디스커션 깜빡 했네요 죄송합니다 ㅠㅠ

Front-line-dev reviewed (approved):
저희 프로젝트에 꼭 필요한 부분을 구현해주시니 샛별님 덕분에 저희 백로그가 쭉쭉 올라가네요. 저녁 늦게까지 수고 많으셨습니다. 정말 우리 팀 열의가 엄청난거 같아요. 아마 샛별님 영향을 받아서 그런거 같네요 ㅋㅋ

jinhyukoo commented on src/backend/utils/number.js:
```diff
@@ -8,4 +8,8 @@ export const CARD = {
+  WAIT_TELLER_SELECT: 6000,
+};
+
+export const WAIT_GAME_START = 60000;
```
이 부분 의도하신건가요...? 제가 아는 WAIT_GAME_START는 대기실에서 게임 들어가는 시간인데... 모르겠다... 🙎‍♂️

sbyeol3 replied:
ㅋㅋ 고치겠습니다

jinhyukoo reviewed (approved):
소켓 util로 분리할 생각을 하시다니! 멋있어요 샛별님 헤헤

sbyeol3 committed:
Refactor : 코드리뷰로 리팩토링

- emit은 socket으로 이동
- time 오타 변경
- force~ 함수는 utils/socket
- select하는 함수는 user로 넣음

sbyeol3 committed:
Refactor : socket 씬으로 이동

- 코드리뷰
